modules/client/src/internal/utils.ts

```diff
@@ -824,3 +863,801 @@ export function toPluginPermissionOperationType(
       throw new InvalidPermissionOperationType();
   }
 }
+
+// function that compares 2 generic arrays
+// and returns true if they are equal
+// and false if they are not
+export function compareArrays<T>(array1: T[], array2: T[]): boolean {
+  if (array1.length !== array2.length) {
+    return false;
+  }
+  for (const [index, element] of array1.entries()) {
+    if (element !== array2[index]) {
+      return false;
+    }
+  }
+  return true;
+}
+async function getPluginInstallations(
+  daoAddress: string,
+  pluginAddress: string,
+  graphql: IClientGraphQLCore,
+): Promise<SubgraphPluginInstallationListItem[]> {
+  const name = "pluginInstallations";
+  type U = { pluginInstallations: SubgraphPluginInstallationListItem[] };
+  const query = QueryPluginInstallations;
+  const params = {
+    where: {
+      plugin: pluginAddress.toLowerCase(),
+      dao: daoAddress.toLowerCase(),
+    },
+  };
+  const res = await graphql.request<U>({
+    query,
+    params,
+    name,
+  });
+  const { pluginInstallations } = res;
+  return pluginInstallations;
+}
+export async function validateGrantUpdatePluginPermissionAction(
+  action: DaoAction,
+  pspAddress: string,
+  daoAddress: string,
+  graphql: IClientGraphQLCore,
+): Promise<PluginUpdateProposalInValidityCause[]> {
+  const causes: PluginUpdateProposalInValidityCause[] = [];
+  const decodedPermission = decodeGrantAction(action.data);
+  const pluginInstallations = await getPluginInstallations(
+    daoAddress,
+    decodedPermission.where,
+    graphql,
+  );
+  if (pluginInstallations.length === 0) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_GRANT_UPDATE_PERMISSION_WHERE_ADDRESS,
+    );
+  }
+  if (action.value.toString() !== "0") {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_GRANT_UPDATE_PERMISSION_VALUE,
+    );
+  }
+  if (decodedPermission.who !== pspAddress) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_GRANT_UPDATE_PERMISSION_WHO_ADDRESS,
+    );
+  }
+  if (
+    decodedPermission.permission !== Permissions.UPGRADE_PLUGIN_PERMISSION
+  ) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_GRANT_UPDATE_PERMISSION_PERMISSION,
+    );
+  }
+  if (
+    decodedPermission.permissionId !==
+      PermissionIds.UPGRADE_PLUGIN_PERMISSION_ID
+  ) {
+    causes.push(
+      PluginUpdateProposalInValidityCause
+        .INVALID_GRANT_UPDATE_PERMISSION_PERMISSION_ID,
+    );
+  }
+  return causes;
+}
+
+export async function validateRevokeUpdatePluginPermissionAction(
```

Comment:
A lot of code duplication just to return the same causes for revoke instead of grant.
We should think about refactoring this in an upcoming task.